tests/basis_test.py

```diff
@@ -1,330 +1,1017 @@
-    assert u1.derivative(-1) == pytest.approx(2)
-    assert u2.derivative(-1) == pytest.approx(-8)
-
-    assert u0.derivative(0.5) == pytest.approx(0)
-    assert u1.derivative(0.5) == pytest.approx(2)
-    assert u2.derivative(0.5) == pytest.approx(4)
-
-    assert numpy.isclose(u0.derivative([1,-1,0.5]),[0,0,0]).all()
-    assert numpy.isclose(u1.derivative([1,-1,0.5]),[2,2,2]).all()
-    assert numpy.isclose(u2.derivative([1,-1,0.5]),[8,-8,4]).all()
-
+import scipy.special
+
+import smolyay
+
+
+def test_cheb_initial():
+    """test degrees returns correctly"""
+    f2 = smolyay.basis.ChebyshevFirstKind(2)
+    assert f2.degree == 2
+    assert isinstance(f2.degree, int)
+    assert numpy.array_equal(f2.domain, [-1, 1])
+    f2.degree = float(3)
+    assert f2.degree == 3
+    assert isinstance(f2.degree, int)
+
+
+@pytest.mark.parametrize(
+    "degree,answer_key",
+    [
+        (0, {0.5: 1, 1: 1, -1: 1, -0.25: 1, -0.5: 1}),
+        (1, {0.5: 0.5, 1: 1, -1: -1, -0.25: -0.25, -0.5: -0.5}),
+        (2, {0.5: -0.5, 1: 1, -1: 1, -0.25: -0.875, -0.5: -0.5}),
+    ],
+)
+def test_cheb_call(degree, answer_key):
+    """Test chebyshev polynomial call"""
+    f = smolyay.basis.ChebyshevFirstKind(degree)
+    assert f(0.5) == pytest.approx(answer_key[0.5])
+    assert f(1) == pytest.approx(answer_key[1])
+    assert f(-1) == pytest.approx(answer_key[-1])
+    assert f(-0.25) == pytest.approx(answer_key[-0.25])
+    assert f(-0.5) == pytest.approx(answer_key[-0.5])
+
+
+@pytest.mark.parametrize(
+    "degree,answer_key",
+    [
+        (0, {0.5: 1, 1: 1, -1: 1, -0.25: 1, -0.5: 1}),
+        (1, {0.5: 0.5, 1: 1, -1: -1, -0.25: -0.25, -0.5: -0.5}),
+        (2, {0.5: -0.5, 1: 1, -1: 1, -0.25: -0.875, -0.5: -0.5}),
+    ],
+)
+def test_cheb_call_1D(degree, answer_key):
+    """Test chebyshev polynomial call with a 1D array"""
+    f = smolyay.basis.ChebyshevFirstKind(degree)
+    get_answer = lambda x: numpy.select(
+        [numpy.array(x) == k for k in answer_key], answer_key.values(), numpy.nan
+    )
+    xs = [0.5, 1, -1, -0.25, -0.5]
+    assert numpy.shape(f(xs)) == numpy.shape(xs)
+    assert numpy.allclose(f(xs), get_answer(xs))
+
+
+@pytest.mark.parametrize(
+    "degree,answer_key",
+    [
+        (0, {0.5: 1, 1: 1, -1: 1, -0.25: 1, -0.5: 1}),
+        (1, {0.5: 0.5, 1: 1, -1: -1, -0.25: -0.25, -0.5: -0.5}),
+        (2, {0.5: -0.5, 1: 1, -1: 1, -0.25: -0.875, -0.5: -0.5}),
+    ],
+)
+def test_cheb_call_2D(degree, answer_key):
+    """Test chebyshev polynomial call with a 2D array"""
+    f = smolyay.basis.ChebyshevFirstKind(degree)
+    get_answer = lambda x: numpy.select(
+        [numpy.array(x) == k for k in answer_key], answer_key.values(), numpy.nan
+    )
+    xs = [[0.5, 1, -1, -0.25], [-0.5, -1, 1, -0.25]]
+    xs2 = numpy.reshape(xs[0], (1, 4))
+    xs3 = numpy.reshape(xs[0], (4, 1))
+    assert numpy.shape(f(xs)) == numpy.shape(xs)
+    assert numpy.allclose(f(xs), get_answer(xs))
+    assert numpy.shape(f(xs2)) == (1, 4)
```

Comment:
Could you please interleave the creation of the various xs and the tests for them, with white space in between? It makes it easier to read if you see the input get created and then immediately tested.